src/pyscheduling/JS/JobShop.py

```diff
@@ -10,22 +10,214 @@
+
+        Args:
+            edges_to_add (list[tuple(tuple(int,int),tuple(int,int))]): list of operations couples where an edge will be added from the first element of a couple to the second element of the couple
+        """
+        adjacency_matrix = dict(self.DG.adjacency())
+
+        #Change the weight of the incident conjunctive edge of the first operation to include setup time
+        first_op = edges_to_add[0][0]
+        precedent_op = [op for op in self.DG.nodes() if self.DG.has_edge(op,first_op)][0]
+        setup_time = self.S[first_op[0]][first_op[1]][first_op[1]] if hasattr(self, 'S') else 0
+        self.DG[precedent_op][first_op]['weight'] = self.DG[precedent_op][first_op]['weight'] + setup_time
+
+        for edge in edges_to_add :
+            processing_time = adjacency_matrix[edge[0]][next(iter(adjacency_matrix[edge[0]]))]['weight']
+            setup_time = self.S[edge[0][0]][edge[0][1]][edge[1][1]] if hasattr(self, 'S') else 0
+            self.DG.add_weighted_edges_from([(edge[0],edge[1],processing_time + setup_time)])
+
+    def generate_precedence_constraints(self, unscheduled_machines : list[int]):
+        precedence_constraints = []
+        for machine_id in unscheduled_machines :
+            vertices = self.get_operations_on_machine(machine_id);
+            for u in vertices :
+                for v in vertices :
+                    if u is not v and nx.has_path(self.DG, u, v) : precedence_constraints.append((u[1],v[1]))
+            
+        return precedence_constraints
+
+    def generate_riPrecLmax(self, machine_id : int, Cmax : int, precedenceConstraints : list[tuple]):
+        """generate an instance of 1|ri,prec|Lmax instance of the machine machine_id
 
+        Args:
+            machine_id (int): id of the machine
+            Cmax (int): current makespan
+
+        Returns:
+            riPrecLmax_Instance: generated 1|ri,prec|Lmax instance
+        """
+        vertices = self.get_operations_on_machine(machine_id)
+        jobs_number = len(vertices)
 
-class GenerationLaw(Enum):
-    UNIFORM = 1
-    NORMAL = 2
+        adjacency_matrix = dict(self.DG.adjacency())
+        P = [-adjacency_matrix[vertice][next(iter(adjacency_matrix[vertice]))]['weight'] for vertice in vertices]
+
+        R = [self.longest_path(self.source,vertice) for vertice in vertices]
 
+        D = [Cmax - self.longest_path(vertices[vertice_ind],self.sink) + P[vertice_ind]
+                                    for vertice_ind in range(len(vertices))]
+
+        return riPrecLmax.riPrecLmax_Instance(name="",n=jobs_number,P=P,R=R,D=D, Precedence=precedenceConstraints)
+
+    def job_completion(self,job_id):
+        """returns the distance of the critical path which corresponds to the Makespan
+
+        Returns:
+            int: critical path distance
+        """
+        return self.longest_path(self.source, self.jobs_sinks[job_id])
+
+    def all_jobs_completion(self):
```

Comment:
This is already implemented in the solution and avoids recomputing the paths when it's not necessary.